CMakeLists.txt

```diff
@@ -23,115 +23,43 @@
 # CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 # TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 # MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+cmake_minimum_required(VERSION 3.14)
+project(SPIRV_HEADERS LANGUAGES CXX VERSION 1.5.5)
 
-#
-# The SPIR-V headers from the SPIR-V Registry
-# https://www.khronos.org/registry/spir-v/
-#
-cmake_minimum_required(VERSION 3.0)
-project(SPIRV-Headers VERSION 1.5.5)
-
-# There are two ways to use this project.
-#
-# Using this source tree directly from a CMake-based project:
-#   1. Add an add_subdirectory directive to include this sub directory.
-#   2. Use ${SPIRV-Headers_SOURCE_DIR}/include} in a target_include_directories
-#      command.
-#
-# Installing the headers first, then using them with an implicit include
-# directory.  To install the headers:
-#   1. mkdir build ; cd build
-#   2. cmake ..
-#   3. cmake --build . --target install
-
-option(SPIRV_HEADERS_SKIP_EXAMPLES "Skip building examples"
-      ${SPIRV_HEADERS_SKIP_EXAMPLES})
-
-option(SPIRV_HEADERS_SKIP_INSTALL "Skip install"
-      ${SPIRV_HEADERS_SKIP_INSTALL})
-
-if(NOT ${SPIRV_HEADERS_SKIP_EXAMPLES})
-  set(SPIRV_HEADERS_ENABLE_EXAMPLES ON)
+if (CMAKE_VERSION VERSION_LESS "3.21")
+    # https://cmake.org/cmake/help/latest/variable/PROJECT_IS_TOP_LEVEL.html
+    string(COMPARE EQUAL ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_SOURCE_DIR} PROJECT_IS_TOP_LEVEL)
 endif()
 
-if(NOT ${SPIRV_HEADERS_SKIP_INSTALL})
-  set(SPIRV_HEADERS_ENABLE_INSTALL ON)
-  # legacy
-  add_custom_target(install-headers
-      COMMAND cmake -E copy_directory ${CMAKE_CURRENT_SOURCE_DIR}/include/spirv
-          $ENV{DESTDIR}${CMAKE_INSTALL_PREFIX}/include/spirv)
-endif()
-
-if (SPIRV_HEADERS_ENABLE_EXAMPLES)
-  message(STATUS "Building SPIRV-Header examples")
-  add_subdirectory(example)
-endif()
+add_library(SPIRV-Headers INTERFACE)
+add_library(SPIRV-Headers::SPIRV-Headers ALIAS SPIRV-Headers)
```

Comment:
I'm not a CMake expert, but would it be better for the alias to be `SPIRV::Headers` rather than `SPIRV-Headers::SPIRV-Headers`?  I could imagine at some point we would have other projects in a `SPIRV` namespace.

Reply:
I chose this name to be consistent with the name already being exported by find_package (See tests/find_package for what I mean).  I agree it's not an ideal name, but it's consistent and will cause less potential confusion.

Reply:
It's this property that was empty for me https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/8b4a4eb257f88e69eadbf30974680bf8cb6d9b84/layers/CMakeLists.txt#L121. I haven't attempted to get to the bottom of it. Providing `SPIRV_HEADERS_INCLUDE_DIR` externally after commenting out this line was good enough for what I was doing but long term I'd like to be able to use upstream unmodified if possible. I'll try again once this goes in and I cycle back to working on this. In the meantime if someone wants to drive this forward, here's the project structure I'm using (simplified):

* My parent CMake project
  * add_subdirectory(SPIRV-Headers)
  * add_subdirectory(Vulkan-Headers)
  * add_subdirectory(Vulkan-ValidationLayers)

In general it'd be great if the various projects did not assume that their dependencies are installed and could rely on existing targets seamlessly. Having this as a tested config upstream would be grand!

Reply:
> It's this property that was empty for me https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/8b4a4eb257f88e69eadbf30974680bf8cb6d9b84/layers/CMakeLists.txt#L121

Currently the vulkan validation layer code assumes it's finding spirv-headers with `find_package` and as you pointed out will break when using `add_subdirectory` instead. There is nothing that `spirv-headers` can do about this.

Please file an issue in the vulkan-validationlayers and I can fix this assumption to assist your workflow.

> In general it'd be great if the various projects did not assume that their dependencies are installed and could rely on existing targets seamlessly.

The problem with using `add_subdirectory` for grabbing dependencies is that it doesn't play well with system/language package managers (vcpkg, conan, etc) which require `flat builds`.

IE solving this problem:

![image](https://user-images.githubusercontent.com/114601453/226998893-61050791-93a7-475f-a1aa-836239a944e1.png)

![image](https://user-images.githubusercontent.com/114601453/226998983-fd715225-84d1-4c99-9054-1c814361a985.png)

Reply:
It may be worth pointing out why VVL focuses on installed dependencies.

Fedora, Debian, Arch Linux, and other distros have vvl as part of their system package management.
- https://salsa.debian.org/xorg-team/vulkan/vulkan-validationlayers/-/blob/debian-unstable/debian/rules

Conan (Language Package manager) does as well.
- https://github.com/conan-io/conan-center-index/tree/master/recipes/vulkan-validationlayers

So they are currently the priority for testing.

Not too mention the SDK packaging for vulkan sdk also utilizes `find_package` / `installation` method due to the aforementioned `flat build` structure benefits that make delivery of multiple libraries/executables significantly simpler / less bug prone.

Apologies this might be more context than needed.

Reply:
Also it's worth pointing out this isn't a CMake exclusive problem. The methodology of building your dependencies as part of your project versus using a package management system is a problem for any C/C++ build system and each build system has their own approach/solution to the problem. And there are pros/cons to either approach.

For now this PR is just a first step towards cleaning things up.

Reply:
Thanks for the explanations, this makes sense to me.  I missed that we previously installed into the `SPIRV-Headers` namespace (the variables were confusing!) and agree it's best to keep this the same.

I guess if we really want to we can add `SPIRV::Headers` alias later, but that's not a problem we need to solve now.

Thanks for cleaning this up!

Reply:
This is causing Chrome and related projects a lot of grief.

We very much do not want to be forced to use the installed version of a package. We want to point to our own specified set of sources.